hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiServiceImpl.java

```diff
@@ -0,0 +1,387 @@
+                    ownerBundleContext.registerService(HazelcastInstance.class.getName(), hazelcastOSGiInstance, null);
+        } else {
+            serviceRegistration = EMPTY_SERVICE_REGISTRATION;
+        }
+        instanceServiceRegistrationMap.put(hazelcastOSGiInstance, serviceRegistration);
+        instanceMap.put(instance.getName(), hazelcastOSGiInstance);
+        return hazelcastOSGiInstance;
+    }
+
+    private void deregisterInstance(HazelcastOSGiInstance hazelcastOSGiInstance) {
+        instanceMap.remove(hazelcastOSGiInstance.getName());
+        ServiceRegistration serviceRegistration =
+                instanceServiceRegistrationMap.remove(hazelcastOSGiInstance);
+        if (serviceRegistration != null && serviceRegistration != EMPTY_SERVICE_REGISTRATION) {
+            ownerBundleContext.ungetService(serviceRegistration.getReference());
+        }
+    }
+
+    private void shutdownAllInternal() {
+        try {
+            for (Map.Entry<HazelcastOSGiInstance, ServiceRegistration> entry : instanceServiceRegistrationMap.entrySet()) {
+                HazelcastOSGiInstance instance = entry.getKey();
+                ServiceRegistration serviceRegistration = entry.getValue();
+                if (serviceRegistration != EMPTY_SERVICE_REGISTRATION) {
+                    try {
+                        ownerBundleContext.ungetService(serviceRegistration.getReference());
+                    } catch (Throwable t) {
+                        LOGGER.finest("Error occurred while deregistering " + instance, t);
+                    }
+                }
+                try {
+                    instance.shutdown();
+                } catch (Throwable t) {
+                    LOGGER.finest("Error occurred while shutting down " + instance, t);
+                }
+                instanceServiceRegistrationMap.remove(instance);
+                instanceMap.remove(instance.getName());
+            }
+        } finally {
+            if (hazelcastInstance != null) {
+                // Default instance may not be registered due to set "HAZELCAST_OSGI_REGISTER_DISABLED" flag,
+                // So be sure that is it shutdown.
+                try {
+                    hazelcastInstance.shutdown();
+                } catch (Throwable t) {
+                    LOGGER.finest("Error occurred while shutting down the default Hazelcast instance !", t);
+                } finally {
+                    hazelcastInstance = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public Bundle getOwnerBundle() {
+        return ownerBundle;
+    }
+
+    @Override
+    public boolean isActive() {
+        return active;
+    }
+
+    @Override
+    public void activate() {
+        // No need to complex lock-free approaches since this is not called frequently. Simple is good.
+        synchronized (serviceMutex) {
+            if (!active) {
```

Comment:
Why do we need an extra active flag here? AFAIK the container should keep track of if a bundle is active or not - we shouldn't need an additional flag.